src/theory/bv/theory_bv_solver.h

```diff
@@ -0,0 +1,130 @@
+/*********************                                                        */
+/*! \file theory_bv.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Liana Hadarean, Andrew Reynolds, Tim King
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Bitvector theory.
+ **
+ ** Bitvector theory.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__BV__THEORY_BV_SOLVER_H
+#define CVC4__THEORY__BV__THEORY_BV_SOLVER_H
+
+#include "theory/bv/theory_bv.h"
+#include "theory/theory.h"
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+class TheoryBVSolver
+{
+ public:
+  TheoryBVSolver(TheoryBV& bv) : d_bv(bv){};
+
+  virtual ~TheoryBVSolver(){};
+
+  /**
+   * Returns true if we need an equality engine. If so, we initialize the
+   * information regarding how it should be setup. For details, see the
+   * documentation in Theory::needsEqualityEngine.
+   */
+  virtual bool needsEqualityEngine(EeSetupInfo& esi) { return false; }
+
+  virtual void finishInit(){};
+
+  virtual void preRegisterTerm(TNode n) = 0;
+
+  virtual void check(Theory::Effort e) = 0;
+
+  virtual bool needsCheckLastEffort() = 0;
+
+  virtual void propagate(Theory::Effort e){};
+
+  virtual TrustNode explain(TNode n)
+  {
+    Unimplemented() << "TheoryBVSolver::explain interface not implemented";
+  };
+
+  virtual bool collectModelInfo(TheoryModel* m) = 0;
+
+  virtual std::string identify() const = 0;
+
+  virtual bool getCurrentSubstitution(int effort,
+                                      std::vector<Node>& vars,
+                                      std::vector<Node>& subs,
+                                      std::map<Node, std::vector<Node>>& exp)
+  {
+    return false;
+  };
+
+  virtual Theory::PPAssertStatus ppAssert(
+      TNode in, SubstitutionMap& outSubstitutions) = 0;
+
+  virtual TrustNode ppRewrite(TNode t) { return TrustNode::null(); };
+
+  virtual void ppStaticLearn(TNode in, NodeBuilder<>& learned){};
+
+  virtual void presolve() = 0;
+
+  virtual void notifySharedTerm(TNode t) = 0;
+
+  virtual EqualityStatus getEqualityStatus(TNode a, TNode b)
+  {
+    return EqualityStatus::EQUALITY_UNKNOWN;
+  }
+
+  virtual Node getModelValue(TNode var) = 0;
+
+  // BV-specific methods
+  virtual bool applyAbstraction(const std::vector<Node>& assertions,
+                                std::vector<Node>& new_assertions) = 0;
+
+ protected:
+  TheoryBV& d_bv;
+
+  // Methods forwarded from TheoryBV
+
+  void computeAssertedTerms(std::set<Node>& termSet,
```

Comment:
This method will be deleted from Theory soon.  I would make this specific to TheoryBVLazy for now.